### PR TITLE
Feature: Block fragments

### PR DESCRIPTION
**Pull request by wrapperup**
Since https://github.com/djc/askama/pull/568 hasn't had an update in a long time, and it is a nice feature, this is a new PR to implement the proposed changes from that discussion.

This adds block fragments/partials as part of the Template derive. Here's an example of how to use it:

```rust
#[derive(Template)]
#[template(path = "...", block = "my_block")]
struct MyBlockFragment<'a> {
    name: &'a str,
}
```

This also properly supports inheritance, nesting, and the `super()` macro. It does this by abstracting `buf_writable` in the generator to add the ability to discard output, except for the block that needs to be rendered.

### CODE REVIEW

File: askama_derive/src/generator.rs
Comment (couchand) [inline askama_derive/src/generator.rs:269]: Two named constructors rather than a `bool` parameter, please.

File: book/src/creating_templates.md
Comment (couchand) [inline book/src/creating_templates.md:78]: This would be clearer if the name of the example block were something other than "block".

File: book/src/template_syntax.md
Comment (couchand) [inline book/src/template_syntax.md:186]: I think it makes sense to mention the feature here since it relates to these blocks, but I feel like it makes sense to put the bulk of the documentation for this in `book/src/creating_templates.md` rather than here.

**Review by couchand — COMMENTED** [review]
I'm really impressed at how effective the strategy of abstracting the `WritableBuffer` is.  Very nice.

Personally this feature seems a bit odd to me.  It feels like its overloading the block inheritance mechanism to do something very different that there's already other ways to do.  I did re-read the thread from #568.  I wonder if perhaps a more complete motivating example would be helpful, in the tests or the docs or just PR comments?

**Comment by wrapperup** [comment]
> Personally this feature seems a bit odd to me. It feels like its overloading the block inheritance mechanism to do something very different that there's already other ways to do.

I couldn't think of a cleaner way here. Perhaps adding another node to do it, but `block` ended up being suggested as the preferred way in that thread.

> I wonder if perhaps a more complete motivating example would be helpful, in the tests or the docs or just PR comments?

It's very much a convenience feature. There are already ways to do achieve the same result (macros, importing other templates). When using a library like `htmx` that send down small sections of html to update the page, it's pretty convenient.

It's part of this whole "locality of behavior" idea, where you keep things as close together as possible (ideally, in the same file). htmx, Alpine.js, Tailwindcss, and all of those that are HTML first trend towards this kind of design. This is also a pretty good read: https://htmx.org/essays/locality-of-behaviour/

File: testing/tests/block_fragments.rs
Comment (djc) [inline testing/tests/block_fragments.rs:8]: Keep the types close to the tests please, instead of keeping types together.

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:392]: Please use a `match` expression for this, and inline it into the initializer expression directly instead of binding here. Let's ditch the constructors for `WritableBuffer` and initialize its fields directly here (and rename the argument to the `Generator::new()` constructor as just `discard` to match).

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:261]: This type should go towards the bottom, just above the `Writable` type definition. I think we should ditch pretty much all the methods and trait impls for it other than the `push()` method, in favor of direct access to its fields.

**Review by djc — COMMENTED** [review]
The direction here seems like it offers a decent amount of functionality especially relative to the limited complexity cost, so it seems reasonable to me. @vallentin / @Kijewski any thoughts?

**Comment by wrapperup** [comment]
The parser crate changes had no effect on this PR. Could I get eyes on these changes again?

**Comment by wsantos** [comment]
@djc what do we need to merge this? I'm considering using it but since my site is going to be built in htmx this is a requirement, thanks in advance.

**Comment by nathanaelhuffman** [comment]
I'm also considering using this with htmx and would love to see this land.

**Comment by pbouzakis** [comment]
Same here; looking to leverage this in htmx project. Currently, I'm using minijinja and render_block, which is working nicely. Would love to move to Askama if I can, but need this feature.

**Comment by teenjuna** [comment]
Hey @Kijewski, sorry to bother, but it looks like @djc is looking for your opinion on this. Can you please take a look? 🙏

**Comment by djc** [comment]
(No is also an okay answer, but would like to have that be made explicit!)

**Comment by Kijewski** [comment]
Sorry, I won't find the time to wrap my head around this PR for the foreseeable time. At a very (_very_) quick glance the code looks fine, and the feature seems useful to me.

**Comment by James-Rhodes** [comment]
Hi all is there any updates on this feature? Very much looking forward to seeing it land :)

**Comment by djc** [comment]
This is waiting for another round of review from me now that we have some positive acknowledgement from @Kijewski (thanks!) -- I'm having a very busy time so it may take me a bit.

(If this is important to you, maybe consider [sponsoring](https://github.com/sponsors/djc) my work on Askama and other projects.)

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
The feature is pretty nice and code looks good to me! Let's wait for @djc to have time to come back to it.

File: askama_derive/src/lib.rs
Comment (djc) [inline askama_derive/src/lib.rs:78]: Given that we're already passing `input` to `Generator::new()`, there's no reason to pass `input.block.is_some()` separately -- it should derive the value for `WritableBuffer::discard`.

**Comment by djc** [comment]
Please just rebase and squash all your changes -- right now "this branch cannot be rebased due to conflicts".

**Comment by djc** [comment]
(Failing CI now.)

**Comment by Kijewski** [comment]
The error is unrelated to this PR, and should be fixed in the current main branch. Can you please rebase, @wrapperup?

**Comment by wrapperup** [comment]
> I think the error is unrelated to this PR, and it should be fixed in the current main branch. Can you please rebase, @wrapperup?

Yep! Was just waiting on it to be fixed in main.

**Comment by djc** [comment]
Thanks, and sorry for the long delays in getting this merged!

**Pull request merged**
